flags/src/main/java/com/yahoo/vespa/flags/file/FlagDirectory.java

```diff
@@ -0,0 +1,156 @@
+                    Optional<FlagData> flagData = readFlagData(flagId);
+                    if (!flagData.isPresent()) return null;
+                    if (!Objects.equals(flagData.get().id(), flagId)) {
+                        logger.log(LogLevel.WARNING, "Flag file " + path + " contains conflicting id " +
+                                flagData.get().id() + ", ignoring flag");
+                        return null;
+                    }
+                    return flagData.get();
+                })
+                .filter(Objects::nonNull)
+                .collect(Collectors.toMap(FlagData::id, Function.identity()));
+    }
+
+    public Optional<FlagData> readFlagData(FlagId flagId) {
+        return readUtf8File(getPathFor(flagId)).map(FlagData::deserialize);
+    }
+
+    /**
+     * Modify the flag directory as necessary, such that a later {@link #read()} will return {@code flagData}.
+     *
+     * @return true if any modifications were done.
+     */
+    public boolean sync(Map<FlagId, FlagData> flagData) {
+        boolean modified = false;
+
+        Set<Path> pathsToDelete = getAllRegularFilesStream().collect(Collectors.toCollection(HashSet::new));
+
+        uncheck(() -> Files.createDirectories(flagDirectory));
+        for (Map.Entry<FlagId, FlagData> entry : flagData.entrySet()) {
+            FlagId flagId = entry.getKey();
+            FlagData data = entry.getValue();
+            Path path = getPathFor(flagId);
+
+            pathsToDelete.remove(path);
+
+            String serializedData = data.serializeToJson();
+            Optional<String> fileContent = readUtf8File(path);
+            if (fileContent.isPresent()) {
+                if (!Objects.equals(fileContent.get(), serializedData)) {
+                    logger.log(LogLevel.INFO, "Updating flag " + flagId + " from " + fileContent.get() +
+                            " to " + serializedData);
+                    writeUtf8File(path, serializedData);
+                    modified = true;
+                }
+            } else {
+                logger.log(LogLevel.INFO, "New flag " + flagId + ": " + serializedData);
+                writeUtf8File(path, serializedData);
+                modified = true;
+            }
+        }
+
+        for (Path path : pathsToDelete) {
+            logger.log(LogLevel.INFO, "Removing flag file " + path);
+            uncheck(() -> Files.deleteIfExists(path));
+            modified = true;
+        }
+
+        return modified;
+    }
+
+    private Stream<Path> getAllRegularFilesStream() {
+        try {
+            return Files.list(flagDirectory).filter(Files::isRegularFile);
+        } catch (NotDirectoryException | NoSuchFileException e) {
+            return Stream.empty();
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private static String getFilenameOf(Path path) {
+        return path.getName(path.getNameCount() - 1).toString();
```

Comment:
Why not `Path::getFileName`?

Reply:
omg, you have no idea how wide I searched for that - I don't know how I missed it.